### ISSUE TITLE
Fine-Tune Open-Weights Code LLM for Java→Bedrock Conversion

### ISSUE THREAD

**Issue by anchapin** [anchapin/ModPorter-AI#997]
## Summary

Once ModPorter has accumulated enough successful conversion examples (~500+), fine-tune an open-weights code LLM to specialize in Java mod → Bedrock add-on translation. This will reduce per-conversion cost, improve domain accuracy, and lower latency compared to API-based models.

## Background

The current pipeline (once #990 is implemented) will use API-based LLMs (GPT-4o / Claude) for code translation. Each successful conversion produces a training pair that can be used to fine-tune a specialized model. This creates a flywheel:

```
Customer submits mod → AI converts → Human reviews/fixes →
Fixed output becomes training data → Model improves →
Next conversion needs less human fixing → Margins improve
```

## Candidate Models (as of April 2026)

| Model | Size | Why It Fits | Fine-Tuning Cost (QLoRA) |
|-------|------|-------------|------------------------|
| **Qwen3-Coder** | 7B–480B | Current SOTA open-source coding model | $50–300/run (7B on RTX 4090) |
| **DeepSeek-Coder-V3** | 16B–236B | Excellent multi-language code understanding; MoE architecture | ~$100–500/run |
| **CodeLlama** | 7B–34B | Meta's code specialist; well-studied fine-tuning pipeline | $50–150/run (7B) |
| **StarCoder2** | 3B–15B | Trained on The Stack v2; good for domain-specific fine-tuning | $50–100/run |

## Recommended Approach

1. **Data Collection (start now):** Design the conversion pipeline to capture training pairs from day one — `(Java source + AST analysis, correct Bedrock output)` for every successful conversion
2. **Fine-Tuning (after 500+ pairs):** Use **QLoRA** on Qwen3-Coder-7B (runs on a single RTX 4090, ~$50–300 per training run)
3. **Validation:** A/B test fine-tuned model vs API model on new conversions
4. **Scale Up:** Move to 32B+ model if 7B hits accuracy ceiling; consider separate models per mod type (blocks, entities, recipes)

## Cost Comparison at Scale

| Approach | Per-Conversion Cost | Quality | Latency |
|----------|-------------------|---------|---------|
| GPT-4o API | ~$0.10–0.50 | High | 5–15s |
| Claude API | ~$0.15–0.75 | High | 5–20s |
| Self-hosted Qwen3-7B (fine-tuned) | ~$0.01–0.03 | Medium→High | 2–5s |
| Self-hosted Qwen3-32B (fine-tuned) | ~$0.03–0.08 | High | 5–10s |

## Future Consideration: Diffusion Language Models

Design the LLM call layer behind an abstraction interface so that diffusion language models (e.g., Mercury by Inception Labs, Dream 7B) can be swapped in later. dLLMs offer 10x speed improvements and native code infilling capabilities, but the ecosystem is still maturing. Watch Mercury's fine-tuning support — a fine-tuned dLLM for Java→Bedrock could be a game-changer.

## Prerequisites

- #990 (LLM-powered AST→NL→Bedrock translation pipeline) must be implemented first
- #991 (RAG pipeline wired into conversion loop) for retrieval-augmented context
- Training data collection mechanism in place

## Acceptance Criteria

- [ ] Training data capture pipeline stores conversion pairs automatically
- [ ] LLM interface is abstracted to support model swapping
- [ ] QLoRA fine-tuning script for Qwen3-Coder-7B (or equivalent)
- [ ] A/B testing framework to compare fine-tuned vs API models
- [ ] Documentation on fine-tuning process and data requirements

## Priority

🔵 **P2 — Future** | Blocked until sufficient training data exists (~500+ conversion pairs)

## References

- [LoRA Fine-Tuning Cost Analysis 2026](https://www.stratagem-systems.com/blog/lora-fine-tuning-cost-analysis-2026)
- [Best Open Source LLMs for Coding 2026](https://www.siliconflow.com/articles/en/best-open-source-LLMs-for-coding)
- [Fine-Tuning SLM vs Prompting LLM (arXiv)](https://arxiv.org/html/2505.24189v1)

**Comment by anchapin** [comment]
## Research Scout Update — 2026-04-18

Adding a research-motivated note to this issue from the weekly PortKit Research Scout sweep.

### Generator-Discriminator RL Signal for Fine-Tuning

New paper: [*RAD-2: Scaling Reinforcement Learning in a Generator-Discriminator Framework*](https://arxiv.org/abs/2604.15308v1) (Gao et al., 2026)

**Relevance to this issue (60%):** When fine-tuning the Java→Bedrock code LLM, we can improve training signal quality by adding a discriminator model trained on pass/fail conversion test results.

**Proposed addition to fine-tuning pipeline:**
- **Generator**: The code LLM produces Bedrock conversion candidates
- **Discriminator**: A smaller model trained on the QA test suite's pass/fail labels scores each candidate
- **RL signal**: Generator is updated using discriminator scores rather than ground-truth-only supervised fine-tuning

This gives two benefits:
1. Training signal is available even when ground-truth conversions are scarce (the discriminator can generalize)
2. The discriminator acts as a proxy for "does this actually work in Bedrock?" without requiring full execution

**Suggested scope addition:** After the base fine-tune is done, consider a second-stage RL pass using a discriminator trained on the QA test library from [#b54408yz7q9lf7ki](https://github.com/anchapin/ModPorter-AI) conversion audit results.

*Flagged by PortKit Research Scout — [full digest](https://github.com/anchapin/ModPorter-AI/blob/main/.planning/research/)*
